feat(search): Adds logic to download search results

ERosendo opened the pull request.

mlissner reviewed (commented):
I gave this a once-over and it feels about right. Concerns I'll highlight for you guys to consider:

1. Memory: We're putting the CSV in memory, which sure is handy. I think this is fine b/c it'll be pretty small, a couple hundred KB, right? This must be fine, but it's on my mind.

3. The fields in the result might be annoying with columns that aren't normalized to human values (like SOURCE: CR or something, and local_path: /recap/gov.xxxx.pdf instead of https://storage.courtlistener.com/recap/gov.xxx.pdf). I didn't see code to fix that, but it's probably something we should do if we can. This CSV is supposed to be for humans, in theory.

I appreciate the refactor, but I'd suggest it in a separate PR in the future, so it's not mixed in.

But this looks about right to me otherwise. :)

albertisfu reviewed (changes requested):
@ERosendo this looks good and is on the right track. I’ve left some comments and suggestions in the code, along with additional feedback here:

- In addition to Mike's comment about normalizing values for humans, a similar suggestion is that I noticed that the CSV headers don’t maintain a fixed order when the CSV is generated and I found it difficult to determine when the results belong to the same "Case," particularly when matching child documents. It might be a good idea to ensure that the headers are fixed for each search type and to prioritize key headers that help identify whether the results belong to the same case. For instance, in RECAP, the headers could start like this:
    
    `docket_id, docket_number, pacer_case_id, court_exact, case_name, document_number, attachment_number, ...`
    
- Highlighted fields in the results are always represented as a list of terms, even though we are only highlighting a single fragment.  Most of the HL fields are not naturally lists of terms. However, some fields, such as citations in case law, can be lists and are also highlighted.
    
    Currently, these fields are rendered as lists:
![Screenshot 2025-01-20 at 6 44 17 p m](https://github.com/user-attachments/assets/f5e34452-63b8-4a3e-8ae8-c7f25c1305bb)

As in the frontend, perhaps you could use the `render_string_or_list` filter or a modified version of it to render HL fields as strings instead of lists when they are not multi-fields?

- Regarding Judge Search, I noticed that the CSV only contains fields from `PersonDocument`, and some fields currently rendered as flat fields in the frontend (such as "Appointers" and other similar fields extracted from the database in `merge_unavailable_fields_on_parent_document`) are not included. Is this behavior expected?

- I'd recommend adding at least one integration test to help catch and prevent regressions in the future related to the suggestions and bugs mentioned in the comments.

Thank you!

ERosendo commented on cl/lib/search_utils.py:
```diff
@@ -261,3 +317,414 @@ def store_search_api_query(
+            keys = AudioDocument.__dict__["_fields"].keys()
+        case SEARCH_TYPES.PARENTHETICAL:
+            keys = ParentheticalGroupDocument.__dict__["_fields"].keys()
+        case SEARCH_TYPES.RECAP:
```
Thank you for bringing this to my attention and for the clarification provided in Slack. You are correct. I will add this search type.

mlissner commented on cl/search/tasks.py:
```diff
@@ -341,6 +352,76 @@ def document_fields_to_update(
+    filename = f'search_results_{now.strftime("%Y%m%d_%H%M%S")}.csv'
+
+    # Send email with attachments
+    message.attach(filename, csv_content, "text/csv")
```
Let's see if this is a problem before we solve it. Good to know about though.

mlissner commented on cl/search/views.py:
```diff
@@ -418,314 +368,11 @@ def es_search(request: HttpRequest) -> HttpResponse:
+@login_required
+@ratelimiter_unsafe_5_per_d
+@require_POST
+def export_search_results(request: AuthenticatedHttpRequest) -> HttpResponse:
```
Sorry, why? I'm imagining a button that triggers this. Shouldn't that work?

albertisfu replied:
It seems that Eduardo is considering using HTMX to trigger this POST request and try to include the CSRF token in the request, so `@csrf_exempt` might not be necessary here

ERosendo commented on cl/lib/search_utils.py:
```diff
@@ -261,3 +317,414 @@ def store_search_api_query(
+        search result.
+    """
+    csv_rows: list[dict[str, Any]] = []
+    while len(csv_rows) <= settings.MAX_SEARCH_RESULTS_EXPORTED:
```
> Could you please explain whether the while loop is required here?

I added the loop bypass the `do_es_search` limitation, which restricts recap document retrieval to `RECAP_SEARCH_PAGE_SIZE` per page, ensuring we get all results up to `MAX_SEARCH_RESULTS_EXPORTED`. 

I'll implement your suggestion to remove the loop

ERosendo commented:
@albertisfu Thanks for your comments. This PR is ready for another review.

Here's a summary with the changes I introduced to address your comments:

-  To address your comment regarding the CSV header order, I've implemented a class method(`get_csv_headers`) that returns a predefined list of keys to use as headers.

- While addressing your suggestion regarding the CSV header order, I found that the ES document classes use a mix of camel case (e.g., [caseName](https://github.com/freelawproject/courtlistener/blob/bbcb637b07e66e596b0058171bf4f8f0b5de8d6a/cl/search/documents.py#L156)) and snake case (e.g., [absolute_url](https://github.com/freelawproject/courtlistener/blob/bbcb637b07e66e596b0058171bf4f8f0b5de8d6a/cl/search/documents.py#L155)) for attribute names. To improve the clarity and consistency of the CSV for users, I've introduced a helper in commit [5f469ca](https://github.com/freelawproject/courtlistener/pull/4893/commits/5f469ca9fbe57ce2d9b073a412eaa0a00651bc35) that converts all keys to snake case.

- As you pointed out, our logic to flatten results was overwriting parent information with child data, which was problematic for highlighted fields. I've resolved this by changing the merge logic. Previously, parent data was the base, and child data was added on top. Now, child data is the base, and parent data is added. This ensures highlights in parent documents are preserved, as they don't conflict with child fields.

- I also added a class method that returns a dictionary with basic data transformations. We'll use this to prepare search results for the CSV file, making the data human-readable and rendering highlighted fields with the `render_string_or_list` helper.

- I've updated the `do_es_search` helper method to add the argument you suggested and got rid of the while loop that was working around the recap search result limit. I also checked if we could use that argument to skip calculating counts, but it would mean changing two other methods. It looks like those ES methods are pretty connected. Maybe we could think about separating them a bit, and then do a separate refactoring to avoid calculating counts later.

Let me know what you think.

albertisfu commented on cl/lib/search_utils.py:
```diff
@@ -261,3 +318,428 @@ def store_search_api_query(
+                | OpinionDocument.get_csv_transformations()
+            )
+
+    return keys, transformations
```
It seems that these variables can be referenced before assignment. Can we just return default empty values for them?

ERosendo replied:
Sure!

ERosendo commented on cl/search/documents.py:
```diff
@@ -1053,12 +1204,68 @@ class ESRECAPBaseDocument(DSLDocument):
+            "rd_page_count",
+            "filepath_local",
+            "rd_plain_text",
+            "cites",
```
Thank you for pointing out the missing document type in the CSV. You're correct; that field is not being included. I'll address this issue.

I've renamed the 'short description' field to `rd_description`. You can review the `new_keys_mapping` variable within the data transformation helper for details.

Also, I've added `attachment_number` and `document_number` to the list of headers returned by the `DocketDocument` class. When I was putting the headers together, it seemed like these worked better at the start of each row.

ERosendo commented on cl/search/tasks.py:
```diff
@@ -351,6 +363,88 @@ def document_fields_to_update(
+    )
+
+    # Create the CSV content and store in a StringIO object
+    csv_content = None
```
This variable is used at line 425

```python
        csv_content: str = output.getvalue()
```

ERosendo commented on cl/search/documents.py:
```diff
@@ -791,14 +891,65 @@ def prepare_political_affiliation_id(self, instance):
+            "date_granularity_dod",
+            "aba_rating",
+            "school",
+            "date_created",
```
Thanks for pointing that out. I'll add these fields.

albertisfu reviewed (approved):
Thanks @ERosendo this looks great now and ready to go!

Just a note from our conversation: When we work on adding the button to the UI for requesting the results to be downloaded, we should implement a validation to ensure the button is only shown for valid searches that return results. This will help avoid triggering the task unnecessarily and retrying the task on queries with syntax errors.

Set to auto-merge!

mlissner commented:
Woo, congrats guys on getting this one done. It was harder than it seemed!
